source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,409 @@
+  ensureMinLength(len, XID_LENGTH + OPCODE_LENGTH + INT_LENGTH);
+  return helper_.peekString(data, offset);
+}
+
+void DecoderImpl::parseCheckRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  ensureMinLength(len, (2 * INT_LENGTH));
+
+  const std::string path = helper_.peekString(data, offset);
+  const int32_t version = helper_.peekInt32(data, offset);
+
+  callbacks_.onCheckRequest(path, version);
+}
+
+void DecoderImpl::parseMultiRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  // Treat empty transactions as a decoding error, there should be at least 1 header.
+  ensureMinLength(len, XID_LENGTH + OPCODE_LENGTH + MULTI_HEADER_LENGTH);
+
+  while (true) {
+    const int32_t opcode = helper_.peekInt32(data, offset);
+    const bool done = helper_.peekBool(data, offset);
+    // Ignore error field.
+    helper_.peekInt32(data, offset);
+
+    if (done) {
+      break;
+    }
+
+    switch (static_cast<OpCodes>(opcode)) {
+    case OpCodes::CREATE:
+      parseCreateRequest(data, offset, len, OpCodes::CREATE);
+      break;
+    case OpCodes::SETDATA:
+      parseSetRequest(data, offset, len);
+      break;
+    case OpCodes::CHECK:
+      parseCheckRequest(data, offset, len);
+      break;
+    default:
+      throw EnvoyException(fmt::format("Unknown opcode within a transaction: {}", opcode));
+    }
+  }
+
+  callbacks_.onMultiRequest();
+}
+
+void DecoderImpl::parseReconfigRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  ensureMinLength(len, XID_LENGTH + OPCODE_LENGTH + (3 * INT_LENGTH) + LONG_LENGTH);
+
+  // Skip joining.
+  skipString(data, offset);
+  // Skip leaving.
+  skipString(data, offset);
+  // Skip new members.
+  skipString(data, offset);
+  // Read config id.
+  helper_.peekInt64(data, offset);
+
+  callbacks_.onReconfigRequest();
+}
+
+void DecoderImpl::parseSetWatchesRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  ensureMinLength(len, XID_LENGTH + OPCODE_LENGTH + (3 * INT_LENGTH));
+
+  // Data watches.
+  skipStrings(data, offset);
+  // Exist watches.
+  skipStrings(data, offset);
+  // Child watches.
+  skipStrings(data, offset);
+
+  callbacks_.onSetWatchesRequest();
+}
+
+void DecoderImpl::parseXWatchesRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len,
+                                       OpCodes opcode) {
+  ensureMinLength(len, XID_LENGTH + OPCODE_LENGTH + (2 * INT_LENGTH));
+
+  const std::string path = helper_.peekString(data, offset);
+  const int32_t type = helper_.peekInt32(data, offset);
+
+  if (opcode == OpCodes::CHECKWATCHES) {
+    callbacks_.onCheckWatchesRequest(path, type);
+  } else {
+    callbacks_.onRemoveWatchesRequest(path, type);
+  }
+}
+
+void DecoderImpl::skipString(Buffer::Instance& data, uint64_t& offset) {
+  const int32_t slen = helper_.peekInt32(data, offset);
+  offset += slen;
+  helper_.add(slen);
```

Comment:
Maybe this should be a `skip` function instead that modifies the offset as well?

Reply:
you mean, skip() within DecoderImpl with just that call?